relay-general/src/protocol/replay.rs

```diff
@@ -0,0 +1,409 @@
+//! They are ingested incrementally.
+//!
+//! # Protocol
+//!
+//! Relay is expecting a JSON object with some mandatory metadata.  However, environment and user
+//! metadata is usually sent in addition to the minimal payload.
+//!
+//! ```json
+//! {
+//!     "type": "replay_event",
+//!     "replay_id": "d2132d31b39445f1938d7e21b6bf0ec4",
+//!     "event_id": "63c5b0f895441a94340183c5f1e74cd4",
+//!     "segment_id": 0,
+//!     "timestamp": 1597976392.6542819,
+//!     "replay_start_timestamp": 1597976392.6542819,
+//!     "urls": ["https://sentry.io"],
+//!     "error_ids": ["d2132d31b39445f1938d7e21b6bf0ec4"],
+//!     "trace_ids": ["63c5b0f895441a94340183c5f1e74cd4"],
+//!     "request": {
+//!         "headers": {"User-Agent": "Mozilla/5.0..."}
+//!     },
+//! }
+//! ```
+use crate::protocol::{
+    ClientSdkInfo, Contexts, IpAddr, LenientString, Request, Tags, Timestamp, User,
+};
+use crate::store::is_valid_platform;
+use crate::store::user_agent::normalize_user_agent_generic;
+use crate::types::{Annotated, Array};
+use crate::user_agent;
+use std::net::IpAddr as RealIPAddr;
+
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+#[metastructure(process_func = "process_replay", value_type = "Replay")]
+pub struct Replay {
+    pub event_id: Annotated<String>,
+    pub replay_id: Annotated<String>,
+    pub replay_type: Annotated<String>,
+    pub segment_id: Annotated<u64>,
+    pub timestamp: Annotated<Timestamp>,
+    pub replay_start_timestamp: Annotated<Timestamp>,
+    pub urls: Annotated<Array<String>>,
+    pub error_ids: Annotated<Array<String>>,
+    pub trace_ids: Annotated<Array<String>>,
+    pub contexts: Annotated<Contexts>,
+    pub platform: Annotated<String>,
+
+    #[metastructure(
+        max_chars = "tag_value",
+        required = "false",
+        trim_whitespace = "true",
+        nonempty = "true",
+        skip_serialization = "empty"
+    )]
+    pub release: Annotated<LenientString>,
+
+    #[metastructure(
+        allow_chars = "a-zA-Z0-9_.-",
+        trim_whitespace = "true",
+        required = "false",
+        nonempty = "true"
+    )]
+    pub dist: Annotated<String>,
+
+    #[metastructure(
+        max_chars = "environment",
+        nonempty = "true",
+        required = "false",
+        trim_whitespace = "true"
+    )]
+    pub environment: Annotated<String>,
+
+    #[metastructure(skip_serialization = "empty", pii = "maybe")]
+    pub tags: Annotated<Tags>,
+
+    #[metastructure(field = "type")]
+    pub ty: Annotated<String>,
+
+    #[metastructure(skip_serialization = "empty")]
+    pub user: Annotated<User>,
+
+    #[metastructure(skip_serialization = "empty")]
+    pub request: Annotated<Request>,
+
+    #[metastructure(field = "sdk")]
+    #[metastructure(skip_serialization = "empty")]
+    pub sdk: Annotated<ClientSdkInfo>,
```

Comment:
Let's put doc comments on top of every `pub` member with a short explanation of what the field is for, like [here](https://github.com/getsentry/relay/blob/4cb923f0667da94801039ff2cbd99635221dd97b/relay-general/src/protocol/event.rs#L174).